src/mbgl/style/conversion/function.cpp

```diff
@@ -0,0 +1,702 @@
+                                                                   Error& error) {
+    auto stopsValue = objectMember(value, "stops");
+    if (!stopsValue) {
+        error = { "function value must specify stops" };
+        return {};
+    }
+
+    if (!isArray(*stopsValue)) {
+        error = { "function stops must be an array" };
+        return {};
+    }
+
+    if (arrayLength(*stopsValue) == 0) {
+        error = { "function must have at least one stop" };
+        return {};
+    }
+
+    std::map<T, std::unique_ptr<Expression>> stops;
+    for (std::size_t i = 0; i < arrayLength(*stopsValue); ++i) {
+        const auto& stopValue = arrayMember(*stopsValue, i);
+
+        if (!isArray(stopValue)) {
+            error = { "function stop must be an array" };
+            return {};
+        }
+
+        if (arrayLength(stopValue) != 2) {
+            error = { "function stop must have two elements" };
+            return {};
+        }
+
+        optional<T> t = convert<T>(arrayMember(stopValue, 0), error);
+        if (!t) {
+            return {};
+        }
+
+        optional<std::unique_ptr<Expression>> e = convertLiteral(type, arrayMember(stopValue, 1), error);
+        if (!e) {
+            return {};
+        }
+
+        stops.emplace(*t, std::move(*e));
+    }
+
+    return { std::move(stops) };
+}
+
+static optional<double> convertBase(const Convertible& value, Error& error) {
+    auto baseValue = objectMember(value, "base");
+
+    if (!baseValue) {
+        return 1.0;
+    }
+
+    auto base = toNumber(*baseValue);
+    if (!base) {
+        error = { "function base must be a number" };
+        return {};
+    }
+
+    return *base;
+}
+
+static std::unique_ptr<Expression> step(type::Type type, std::unique_ptr<Expression> input, std::map<double, std::unique_ptr<Expression>> stops) {
+    return std::make_unique<Step>(type, std::move(input), std::move(stops));
+}
+
+static std::unique_ptr<Expression> interpolate(type::Type type, Interpolator interpolator, std::unique_ptr<Expression> input, std::map<double, std::unique_ptr<Expression>> stops) {
+    ParsingContext ctx;
+    auto result = createInterpolate(type, std::move(interpolator), std::move(input), std::move(stops), ctx);
+    if (!result) {
+        assert(false);
+        return {};
+    }
+    return std::move(*result);
+}
+
+template <class T>
+std::unique_ptr<Expression> match(type::Type type, const std::string& property, std::map<T, std::unique_ptr<Expression>> branches) {
+    std::unordered_map<T, std::shared_ptr<Expression>> convertedBranches;
+    for (auto& b : branches) {
+        convertedBranches[b.first] = std::move(b.second);
+    }
+    return std::make_unique<Match<T>>(type, get(literal(property)), std::move(convertedBranches), error("replaced with default"));
+}
+
+template <>
+std::unique_ptr<Expression> match<bool>(type::Type type, const std::string& property, std::map<bool, std::unique_ptr<Expression>> branches) {
```

Comment:
I think it would be slightly clearer to call this `case()` instead of `match()`

Reply:
`case` is a keyword though. I renamed to `categorical`.